### DISCUSSION TITLE
Can you help me understand what imgui_bundle is?

### DISCUSSION THREAD

**Discussion by YAMLcase** [pthom/imgui_bundle#214] [Q&A]
Long time python developer here, looking to get into GUI app development.  My journey for my next project has brought me to immediate mode and am building a list of libraries to use.  So far I've found dear pygui, pyimgui and now this one imgui_bundle. 

What is imgui_bundle?  is it a python library providing the bindings to imgui?  Your intro page says I can also write in C++ which is not something I'd normally find with a python library.  Thanks

**Comment by pthom** [comment]
Hello, nice to hear from you, and welcome to Dear ImGui Bundle!

Dear ImGui, this is a C++ library.  Dear ImGui Bundle provides an access to Dear ImGui, and many of its sister libraries for *C++ **and** Python* developers.
Here is a Hello World in 7 lines in python and C++:

<img width="1001" alt="image" src="https://github.com/pthom/imgui_bundle/assets/7694091/84aa9f9d-8e76-491c-a0f7-3e60e9b2c62d">
 
Please refer to the interactive manual: https://traineq.org/ImGuiBundle/emscripten/bin/demo_imgui_bundle.html
For a quick access to demos with code, click on "Immediate Apps": 
<img width="133" alt="image" src="https://github.com/pthom/imgui_bundle/assets/7694091/80675205-5afb-4868-970d-32d8ba98144f">


* *When compared to pyimgui & dearpygui:*
  * It offers up-to-date and auto-generated python bindings to ImGui so that you can use the latest features of Dear ImGui (the bindings for those two libraries are manual, at least for pyimgui)
  * It offers access to many more sister libraries. pyimgui does not include any, DearPyGui offers a few, but ImGui offers [many more](https://pthom.github.io/imgui_bundle/introduction.html)
  
* *When compared to DearPyGui:*
   * It does respect the immediate mode paradigm, when DearPyGui kind of transformed the immediate mode into a more retained mode.


------------


Now it is my turn to ask you some questions: since you asked this question, it means that, the introduction was not clear enough. 

So I would like some feedback from you to try to improve it in the future. 

- ﻿I suppose this part of the introduction was not clear enough for you. 

> Dear ImGui Bundle is a comprehensive bundle for Dear ImGui, featuring various powerful libraries from its ecosystem. Designed to facilitate the creation of applications in C++ and Python across Windows, macOS, Linux, iOS, Android, and emscripten (Web apps), it is ideal for application developers, and researchers eager to dive into GUI development with ease and efficiency. This bundle simplifies the process, allowing you to focus on the creative aspects of your projects.

What would have made it clearer for you? Maybe by adding a screenshot of a program such as the hello word that I include here ?

- Did you find the acces to the interactive demonstration which can be accessed by clicking on the image below in the introduction?
<img width="315" alt="image" src="https://github.com/pthom/imgui_bundle/assets/7694091/5f4ac5ad-3173-4d06-8c0c-5aec944119f4">

- If you accessed the interactive demonstration, did you access the demonstrations with code in the immediate apps tab, or is it too hidden in the large top bar?

**Reply by YAMLcase** [reply]
Hi Pthom, thank you for the detailed reply!  I'd be happy to help.

First, let me know if I'm on the right track, correct me where I'm wrong :) 

I think what's tripping me up most of all is the word "bundle".  I don't know what a bundle is in the context of what I'm looking for, a python library to create immediate mode GUI applications.  It took me a few times reading through it to gather this bundle is several libraries, one of imgui, and some more for add-ons (node editor, implot, etc).  Another point that kinda threw me off from groking it earlier is "bundle" and "batteries included" used in the same context tricking my brain into thinking they're two separate things.

Another point of confusion is the ability to "create applications in C++ and Python".  I'm looking for a python library, I don't expect to be able to use that same library for other languages.  I'm still not clear on how the two languages seem held at the same level.

I did go through the examples page and am very impressed, but I'm not clear on where `imgui` and `imgui_bundle` showcases differ.  Also, I was expecting to use it in lieu of standing up a hello world on my own.  I will make a hello world later today just to see if I can fill in the blanks for myself.

Final thought: If this project's intent is to be an Immediate Mode GUI toolkit/library/whatever for people to develop in python without having to touch C++, I hope to see it succeed!  Dear PyGUI is in maintenance mode while the author goes off to write Dear PyGUI 2.0 as a complete re-write.  This is the kiss of death as far as I'm concerned.  And pyimgui _sounds like_ a scaled down version of your project.

**Comment by pthom** [comment]
Many thanks for this detailed answer. I will take it into account order to try and make it easier to grasp. 

> I think what's tripping me up most of all is the word "bundle". I don't know what a bundle is

Yes, by bundle I mean that it includes several libraries including ImGui. But you're right, this is not a good name, but it is really too late to change it. 

> Another point of confusion is the ability to "create applications in C++ and Python". I'm looking for a python library, I don't expect to be able to use that same library for other languages. I'm still not clear on how the two languages seem held at the same level.

Yes, I should separate the introduction for C++ and Python developers. The library offers almost the same services for both of them, But there maybe should be a different landing page or part for Python and for C++ developers.

>  but I'm not clear on where imgui and imgui_bundle showcases differ

Actually, the bundle is not a library in itself, It is just a container for many other libraries. 

For example, you would write:

```python
from imgui_bundle import imgui, implot, hello_imgui
```
and then use those libraries. 
   
> I did go through the examples page and am very impressed

Thank you!

> Final thought: If this project's intent is to be an Immediate Mode GUI toolkit/library/whatever for people to develop in python without having to touch C++, I hope to see it succeed! Dear PyGUI is in maintenance mode while the author goes off to write Dear PyGUI 2.0 as a complete re-write. This is the kiss of death as far as I'm concerned. And pyimgui sounds like a scaled down version of your project.

Actually many binding libraries maintainers get tired because they continuously have to update the bindings manually. ImGui Bundle does not have this problem because it generates bindings automatically. This is the reason why it can also include so many other libraries.

**Comment by pthom** [comment]
I updated the [documentation site](https://pthom.github.io/imgui_bundle/)

If you have time, please tell me if you think it is clearer now.
